public/jsx/components/card-form.jsx

```diff
@@ -138,135 +110,122 @@ module.exports = React.createClass({
-        if (detectedCard && that.cardPatterns[detectedCard]) {
-           cardData = utils.defaults(
-                        that.cardPatterns[detectedCard][validator], cardData);
-        }
-        pattern = cardData.pattern;
-        placeholder = cardData.placeholder;
-        label = cardData.label || cardData.placeholder;
-      }
+  updatePattern: function(fieldId, cardType) {
+    // Update the pattern for card + cvv field if card was detected.
+    if (cardType && this.cardPatterns[cardType]) {
+      return utils.defaults(
+        this.cardPatterns[cardType][fieldId] || {},
+        this.cardPatterns.default[fieldId]
+      );
+    } else {
+      return this.cardPatterns.default[fieldId];
+    }
+  },
 
-      // For non card-validator fields provide a fallback for label;
-      if (!validator) {
-        label = field.label || field.placeholder;
-      }
+  render: function() {
+    var formIsValid = true;
+    var { card, expiration, cvv, ...formAttrs } = this.props;
+
+    // Initial Field Props
+    var fieldProps = {
+      card: card,
+      expiration: expiration,
+      cvv: cvv,
+    };
 
-      if (fieldClassNames.length) {
-        fieldClass = fieldClassNames.join(' ');
+    var that = this;
+    // Iterating over the fieldProps augment the initial
+    // props with dynamic updates.
+    Object.keys(fieldProps).forEach(function(key) {
+      var val = that.stripPlaceholder(that.state[key]);
+      var fieldData = fieldProps[key];
+      var valData = fieldData.validator(val);
+      var cardData = valData.card || {};
+      var id = fieldData.id;
+      fieldData.type = fieldData.type || 'text';
+      fieldData.hasVal = val.length > 0 || false;
+      fieldData.isValid = valData.isValid === true;
+      fieldData.showError = !valData.isValid && !valData.isPotentiallyValid;
+      fieldData.cardType = cardData.type ? cardData.type : null;
+      fieldData.onChangeHandler = that.handleChange;
+
+      if (fieldData.showError === true || fieldData.hasVal === false) {
+        formIsValid = false;
       }
 
-      // We're assuming all fields are required.
-      // so we need to mark the form as invalid
-      // if the value of this field is an empty string.
-      // TODO: consider using the required attr.
-      if (val === '') {
-        allValid = false;
+      // Handle Card Pattern changes.
+      // TODO: there's a bug in the InputMask that means
+      // pattern changes are no-op.
+      var patternData = that.updatePattern(id, fieldData.cardType);
+      fieldData.pattern = patternData.pattern;
+      fieldData.placeholder = patternData.placeholder;
+      fieldData.label = patternData.label || patternData.placeholder;
+
+      // Handle field level API errors.
+      if (that.state.errors.fields && that.state.errors.fields[id]) {
+        fieldData.isValid = false;
+        fieldData.showError = true;
+        fieldData.errorMessage = that.state.errors.fields[id][0].message;
+        formIsValid = false;
```

Comment:
it feels like a lot of ``fieldData`` logic in here could be moved to render() of the CardInput component, no?